File: mobile_app/lib/domain/entities/sensor_reading.dart
import 'package:equatable/equatable.dart';

class SensorReading extends Equatable {
  const SensorReading({
    required this.fieldId,
    required this.ph,
    required this.nitrogen,
    required this.phosphorus,
    required this.potassium,
    required this.moisture,
    required this.temperature,
    this.humidity,
    required this.timestamp,
  });

  final String fieldId;
  final double ph;
  final double nitrogen;
  final double phosphorus;
  final double potassium;
  final double moisture;
  final double temperature;
  final double? humidity; // ⭐ DOUBLE not String
  final DateTime timestamp;

  @override
  List get props => [
        fieldId,
        ph,
        nitrogen,
        phosphorus,
        potassium,
        moisture,
        temperature,
        humidity,
        timestamp,
      ];

  factory SensorReading.fromJson(Map json) {
    return SensorReading(
      fieldId: json['field_id'] as String? ?? 'unknown',
      ph: (json['ph'] as num?)?.toDouble() ?? 0.0,
      nitrogen: (json['nitrogen'] as num?)?.toDouble() ?? 0.0,
      phosphorus: (json['phosphorus'] as num?)?.toDouble() ?? 0.0,
      potassium: (json['potassium'] as num?)?.toDouble() ?? 0.0,
      moisture: (json['moisture'] as num?)?.toDouble() ?? 0.0,
      temperature: (json['temperature'] as num?)?.toDouble() ?? 0.0,
      humidity: (json['humidity'] as num?)?.toDouble(), // ⭐ CONVERT to double
      timestamp: DateTime.parse(json['timestamp'] as String),
    );
  }
}

File: backend/app/models/sensor_reading.py
"""Sensor reading data model"""
from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, Optional


class SensorReading:
    """Sensor reading model for IoT data"""

    def __init__(
        self,
        reading_id: str,
        field_id: str,
        device_id: str,
        ph: Optional[float] = None,
        nitrogen: Optional[float] = None,
        phosphorus: Optional[float] = None,
        potassium: Optional[float] = None,
        moisture: Optional[float] = None,
        temperature: Optional[float] = None,
        humidity: Optional[float] = None,
        timestamp: Optional[datetime] = None,
        **kwargs: Any,
    ) -> None:
        self.reading_id = reading_id
        self.field_id = field_id
        self.device_id = device_id
        self.ph = ph
        self.nitrogen = nitrogen
        self.phosphorus = phosphorus
        self.potassium = potassium

        self.moisture = moisture
        self.temperature = temperature
        self.humidity = humidity
        self.timestamp = timestamp or datetime.utcnow()

        for key, value in kwargs.items():
            setattr(self, key, value)

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for Firestore"""
        return {
            'reading_id': self.reading_id,
            'field_id': self.field_id,
            'device_id': self.device_id,
            'ph': self.ph,
            'nitrogen': self.nitrogen,
            'phosphorus': self.phosphorus,
            'potassium': self.potassium,
            'moisture': self.moisture,
            'temperature': self.temperature,
            'humidity': self.humidity,
            'timestamp': self.timestamp
        }

    @staticmethod
    def from_dict(data: Dict[str, Any]) -> 'SensorReading':
        """Create SensorReading from Firestore document"""
        return SensorReading(**data)

    def __repr__(self) -> str:
        return f"<SensorReading {self.reading_id} - Field: {self.field_id}>"
